packages/contracts/contracts/governance/MintManager.sol

```diff
@@ -0,0 +1,136 @@
+
+    /**
+     * @notice The denominator of each recipient's share.
+     */
+    uint256 public constant SHARE_DENOMINATOR = 100;
+
+    /**
+     * @notice The GovernanceToken that the MintManager can mint.
+     */
+    GovernanceToken public immutable GOVERNANCE_TOKEN;
+
+    /**
+     * @notice True when already minted on this chain. MintManager can mint only once on each chain.
+     */
+    bool public minted;
+
+    /**
+     * @notice A list of recipient addresses that will receive tokens to be distributed.
+     */
+    address[] public recipients;
+
+    /**
+     * @notice A mapping of the recipient's address to share.
+     */
+    mapping(address => uint256) public shareOf;
+
+    /**
+     * @notice Constructs the MintManager contract.
+     *
+     * @param _governanceToken The GovernanceToken this contract can mint tokens of.
+     * @param _owner           The owner of this contract.
+     * @param _recipients      List of the recipients.
+     * @param _shares          List of token distribution ratios for each recipient.
+     */
+    constructor(
+        address _governanceToken,
+        address _owner,
+        address[] memory _recipients,
+        uint256[] memory _shares
+    ) {
+        GOVERNANCE_TOKEN = GovernanceToken(_governanceToken);
+
+        transferOwnership(_owner);
+
+        require(_recipients.length == _shares.length, "MintManager: invalid length of array");
+
+        uint256 totalShares = 0;
+        for (uint256 i = 0; i < _recipients.length; i++) {
+            address recipient = _recipients[i];
+            require(recipient != address(0), "MintManager: recipient address cannot be 0");
+
+            uint256 share = _shares[i];
+            require(share != 0, "MintManager: share cannot be 0");
+
+            totalShares += share;
+            recipients.push(recipient);
+            shareOf[recipient] = share;
+        }
+
+        require(
+            totalShares <= SHARE_DENOMINATOR,
+            "MintManager: max total share is equal or less than SHARE_DENOMINATOR"
+        );
+    }
+
+    /**
+     * @notice Only the owner is allowed to mint mint cap amount of the GovernanceToken at once.
+     */
+    function mint() external onlyOwner {
+        require(!minted, "MintManager: already minted on this chain");
+
+        uint256 mintCap = MINT_CAP * 10 ** GOVERNANCE_TOKEN.decimals();
+
+        uint256 totalAmount;
+        for (uint256 i = 0; i < recipients.length; i++) {
+            address recipient = recipients[i];
+            uint256 share = shareOf[recipient];
+            uint256 amount = (mintCap * share) / SHARE_DENOMINATOR;
+            totalAmount += amount;
+        }
+
+        GOVERNANCE_TOKEN.mint(address(this), totalAmount);
+
+        minted = true;
+    }
+
+    /**
+     * @notice Only the owner is allowed to distribute the GovernanceToken to specified recipients.
+     */
+    function distribute() external onlyOwner {
```

Comment:
The `distribute()` should also be called only once. Do we need to check if it's already distributed or not like `minted`? I don't think it's necessary because once distributed, the balance is 0, but I'm asking just in case.

Reply:
Yes, as you said, the call before minting will be execution reverted because the token balance of this contract is 0. Also in the case of call after distribution will be execution reverted.